automap.c

```diff
@@ -386,256 +642,968 @@ famv_isdisjoint(FAMVObject *self, PyObject *other)
+                case KAT_FLOAT16:
+                    k = npy_half_to_double(*(npy_half*)PyArray_GETPTR1(a, table[table_pos].keys_pos));
+                    break;
+                default:
+                    return -1;
+            }
+            if (key == k) {
+                return table_pos;
+            }
+            table_pos++;
+        }
+        table_pos = (5 * (table_pos - SCAN) + (mixin >>= 1) + 1) & mask;
+    }
+}
+
+
+// Compare a passed Py_UCS4 array to stored keys. This does not use any dynamic memory. Returns -1 on error.
+static Py_ssize_t
+lookup_hash_unicode(
+        FAMObject *self,
+        Py_UCS4* key,
+        Py_ssize_t key_size,
+        Py_hash_t hash)
+{
+    TableElement *table = self->table;
+    Py_ssize_t mask = self->table_size - 1;
+    Py_hash_t mixin = Py_ABS(hash);
+    Py_ssize_t table_pos = hash & mask;
+
+    PyArrayObject *a = (PyArrayObject *)self->keys;
+    // REVIEW: is this a new descr reference?
+    Py_ssize_t dt_size = PyArray_DESCR(a)->elsize / sizeof(Py_UCS4);
+
+    Py_hash_t h = 0;
+    Py_UCS4* p_start = NULL;
+
+    while (1) {
+        for (Py_ssize_t i = 0; i < SCAN; i++) {
+            h = table[table_pos].hash;
+            if (h == -1) {
+                return table_pos;
+            }
+            if (h != hash) {
+                table_pos++;
+                continue;
+            }
+            p_start = (Py_UCS4*)PyArray_GETPTR1(a, table[table_pos].keys_pos);
+            // memcmp returns 0 on match
+            if (!memcmp(p_start, key, (key_size < dt_size ? key_size : dt_size))) {
+                return table_pos;
+            }
+            table_pos++;
+        }
+        table_pos = (5 * (table_pos - SCAN) + (mixin >>= 1) + 1) & mask;
+    }
+}
+
+
+// Compare a passed char array to stored keys. This does not use any dynamic memory. Returns -1 on error.
+static Py_ssize_t
+lookup_hash_string(
+        FAMObject *self,
+        char* key,
+        Py_ssize_t key_size,
+        Py_hash_t hash)
+{
+    TableElement *table = self->table;
+    Py_ssize_t mask = self->table_size - 1;
+    Py_hash_t mixin = Py_ABS(hash);
+    Py_ssize_t table_pos = hash & mask;
+
+    PyArrayObject *a = (PyArrayObject *)self->keys;
+    Py_ssize_t dt_size = PyArray_DESCR(a)->elsize / sizeof(char);
+
+    Py_hash_t h = 0;
+    char* p_start = NULL;
+
+    while (1) {
+        for (Py_ssize_t i = 0; i < SCAN; i++) {
+            h = table[table_pos].hash;
+            if (h == -1) {
+                return table_pos;
+            }
+            if (h != hash) {
+                table_pos++;
+                continue;
+            }
+            p_start = (char*)PyArray_GETPTR1(a, table[table_pos].keys_pos);
+            // memcmp returns 0 on match
+            if (!memcmp(p_start, key, (key_size < dt_size ? key_size : dt_size))) {
+                return table_pos;
+            }
+            table_pos++;
+        }
+        table_pos = (5 * (table_pos - SCAN) + (mixin >>= 1) + 1) & mask;
```

Comment:
This is correct. As it's shifted down, `mixin` will start including higher bits of the hash that were ignored during the original index calculation. It's a way of introducing entropy for hashes that collide in the low bits but not the higher ones.

If it truly gets shifted down to zero (if we get *very* unlucky with our table layout and hit over 1,000 colliding items), it means that we have no more entropy in the hash left to use.  If that's the case, then using zeroes for the `mixin` value is (basically) just as useful as using any other bit pattern.

**More importantly, though, it's also required for correctness that `mixin` eventually becomes zero!** The sequence `i = (5 * i + some_random_number + 1) % tablesize` is *not* guaranteed to visit every spot in a table with a size that's a power of two (and it may even visit some spots many times). However, `i = (5 * i + 1) % tablesize` is (and each spot will be visited *exactly* once)! By ensuring that `mixin` eventually becomes (and stays) zero, we can guarantee that all lookups terminate eventually (in the worst case, by visiting every single occupied slot in the table before finally hitting an empty one).

Reply:
[Further reading.](https://github.com/python/cpython/blob/main/Objects/dictobject.c#L165-L217)

Reply:
That makes a lot of sense - thanks! :)

Reply:
Thanks, @brandtbucher , for explaining this!